### ISSUE TITLE
Passing a context into `WebhookService.Parse`.

### ISSUE THREAD

**Issue by chhsia0** [drone/go-scm#43]
In some drivers, e.g., github, we might need a context for its webhook service to reach out to the API endpoint to retrieve additional information. For example, when processing an issue comment event from a pull request, we might need to issue another http request to get the pull request information which is not in the event payload.

So I'm suggesting to change the signature of the Parse function to:
```go
Parse(ctx context.Context, req *http.Request, fn SecretFunc) (Webhook, error)
```
Since this is a breaking change, I'd like to raise some discussion first.

**Comment by chhsia0** [comment]
@bradrydzewski Could you provide some feedback about this? Thanks.

**Comment by chhsia0** [comment]
An alternative to avoid the breaking change is to use `req.Context()` as the context to reach out to the API endpoint. However that means if the caller want to interrupt the call, they needs to set the context object in the request first and use it to control the Parse call, so it's less obvious. WDYT?

**Comment by bradrydzewski** [comment]
I really dislike the way we parse webhooks today.  I would like to refactor to more closely match the go-github library where parse and verify are separated into different methods, and the callback removed.

I believe this can be done by adding new methods, instead of breaking existing methods.

> For example, when processing an issue comment event from a pull request, we might need to issue another http request to get the pull request information which is not in the event payload.

If you need to augment the webhook with additional information, I am of the opinion that this should be handled separately, outside of the library.  For example, in Drone if the commit sha is empty for a tag creation webhook, we make an API call to retrieve additional information about the tag to fill in the blanks.

Also, speaking in the context of Drone, when we parse a webhook we do not know the repository name yet, nor do we know who owns the repository. If we cannot associate the webhook with a repository and, more importantly, a user, we do not have an oauth2 token that we can use to make additional API calls.

**Comment by chhsia0** [comment]
The idea of a `normalize` package is neat. However, either the library requires the user to always use the package, or the webhook struct needs to have some indicator for the user to know when to call it.

For the latter, I just came up with a couple options:
1. Change `PullRequestHook` (which seems okay because it is not supported by and service yet):
```go
type PullRequestHook struct {
    ...
    PullRequest       *PullRequest
    PullRequestNumber int
}
```
Or
```go
type PullRequestHook struct {
    ...
    PullRequest PullRequest
    Normalized  bool
}
```
2. Add a `Normalized` field in `PullRequest`:
```go
type PullRequest struct {
    Number     int
    Normalized bool
    ...
}
```
3. Add an `IsNormalized()` function to the `WebhookService` interface, so each driver does its own check.
4. Add an `IsNormalized()` receiver function to `Webhook` or `PullRequest` that do the same check for all drivers (e.g., check if `Sha` is set).
5. Add another function that returns (*scm.Webhook, bool, error)` to the `WebhookService` interface, where the second return value indicates if the webhook is normalized. If `false` is returned, then only certain documented fields in a particular webhook struct is valid (e.g., `PullRequestCommentWebhook.PullRequest.Number`). This could be made to align with go-github.

After listing all possibilities, I'm now a bit leaning towards option 5. WDYT?

**Comment by bradrydzewski** [comment]
> the library requires the user to always use the package

I prefer that the user always invokes the normalizer package if needed.  Most people will never need this package, but if they do, it can be opt-in.

**Comment by chhsia0** [comment]
What I mean is that, say `PullRequestCommentWebhook`s from github always requires a normalization, but not gitlab. Since the user shouldn't care about what's the underlying driver, if we present this normalization as a utility function, that means the user must always opt-in, or they will get an incomplete `PullRequestCommentWebhook`.

**Comment by bradrydzewski** [comment]
>  if we present this normalization as a utility function, that means the user must always opt-in

Correct, we would require a user to always opt-in if they want to receive the pull request comment webhook for GitHub. This will not impact any of our existing users since this webhook is currently a noop.

I understand where you are coming from, but I think this just comes down to personal preference. I am not sure what sort of long term solution I want. I am not sure I understand the scope of this problem well enough to design that solution now. So instead I would opt for a short term solution that does not make any breaking changes, nor does it introduce changes to the core API that we may need to break in the future.
